src/lsp4clj/server.clj

```diff
@@ -351,9 +351,16 @@
         (if-let [{:keys [p started] :as req} (get pending-requests id)]
           (do
             (trace this trace/received-response req resp started now)
-            (if error
-              (p/reject! p (ex-info "Received error response" resp))
-              (p/resolve! p result)))
+            ;; Note that we are called from the server's pipeline, a core.async
+            ;; go-loop, and therefore must not block.  Callbacks of the pending
+            ;; request's promise will be executed in the completing thread,
+            ;; which should not be our thread.  This is very easy for users to
+            ;; miss, therefore we complete the promise on the default executor.
+            (p/thread-call :default
```

Comment:
> Is it possible for users to be using the default executor as well?

Not sure if I understand what you mean. Of course they can use the default executor, but there are caveats:

- they would need to remember to pass it to every `p/then` (and friends) on the returned promise
- last time I checked there was no `p/catch` that accepts an executor

It does not matter which thread executes the original `send-request` call, only which thread completes the promise.

> And are we sure that our thread isn't in the default executor?

Yes, as we are called from the `pipeline`, which is a `core.async/go-loop`. Our thread is therefore in core.async's fixed-size `dispatch` thread-pool, that executes all `go` blocks.

Reply:
> Our thread is therefore in core.async's fixed-size dispatch thread-pool,

Got it. That makes sense. 

> Not sure if I understand what you mean

I guess my question was whether we're causing problems for our users, by having them to execute their callbacks in the default executor (the common fork-join pool). If they already have a lot of stuff going on in that executor, could it lead to other issues for them? If so, what would we do? Is this materially better or worse for them than what we had before? I lack the familiarity with the underlying tools to answer those questions.

Reply:
I think it is definitely worse, since right now we run code that can potentially block in a thread pool that is not designed for this. In general, I think `go` blocks are really only meant to move things between channels, everything that requires non-trivial computation or even I/O should be done in threads.

But sure, it would make sense to allow users to specify the executor for responses. I will update this PR to make it configurable.